pkg/core/resource_graph.go

```diff
@@ -235,3 +259,214 @@ func (rg *ResourceGraph) getAllDownstreamResourcesSet(source Resource, upstreams
 	}
 	return upstreams
 }
+
+// CreateDependencies takes in a resource and set of metadata and looks at any fields which point to specific dependencies which match the Resource Interface
+// If a specific dependency is found, the .Create method will be called to create the parent resource.
+// Each specific dependency value will be updated on the resources field itself
+// This method will recurse down each one of the resources fields and do a DFS until a Resource is found
+//
+// Before adding a dependency and setting the value on the resource's field, the method will ensure there is no existing node in the graph for the named node created
+// If there is an existing node, the method will ensure the resource's field is set to point to the already existent node
+//
+// IaCValues today are not labeled as a specific resource, so the value must already be present. This method will still add the dependency and ensure the resource is created based on the params passed in, but
+// does not have the knowledge of which resource to create
+//
+//	T params are a map of the reflection field names, to their respective fields
+//
+// Example:
+// for a struct:
+//
+//	 Test{
+//	   Field1 Resource
+//	   Field2 Resource
+//	}
+//
+// The corresponding params would be:
+//
+//	params := map[string]any{
+//	   "Field1": ResourceParams
+//	   "Field2": ResourceParams
+//	}
+//
+// Params for direct resources or IaCValues correlate to the field they are for
+// For params on fields which correlate to the following types, the formats are:
+//   - Map: map[string]ParamType    (the string key corresponds to the key in the map)
+//   - Struct: map[string]ParamType (the string key corresponds to the field in the struct)
+//   - Array or slice: []ParamType
+func (rg *ResourceGraph) CreateDependencies(res Resource, params map[string]any) error {
+	var merr multierr.Error
+	source := reflect.ValueOf(res)
+
+	for source.Kind() == reflect.Pointer {
+		source = source.Elem()
+	}
+	for i := 0; i < source.NumField(); i++ {
+		targetValue := source.Field(i)
+		fieldsParams := params[source.Type().Field(i).Name]
+		if fieldsParams != nil {
+			merr.Append(rg.actOnValue(targetValue, res, fieldsParams, nil, reflect.Value{}))
+		}
+	}
+	rg.AddDependenciesReflect(res)
+	return merr.ErrOrNil()
+}
+
+// CreateResource is a wrapper around a Resources .Create method
+//
+// CreateResource provides safety in assuring that the up to date resource (which is inline with what exists in the ResourceGraph) is returned
+func CreateResource[T Resource](rg *ResourceGraph, params any) (resource T, err error) {
+
+	res := reflect.New(reflect.TypeOf(resource).Elem()).Interface()
```

Comment:
Do you need to use reflection to create an instance of T here? Isn't `resource` already a zero value of type `T`?

Reply:
The zero value would be `nil`, but we can add another type parameter to make it work simpler:

```go
func CreateResource[T interface{*IT; Resource}, IT any](rg *ResourceGraph, params any) (resource T, err error) {

	res := new(IT)
```
This says that "T is a Resource and a pointer to IT", then you can use `new(IT)` to make a new, non-nil T (that is, a `*IT`)

Reply:
oh i can try that out. Yeah to gordons point it is nil, but if the other method works then ill go with that

Reply:
by going out this route i see `invalid operation: res (variable of type *IT) is not an interface` when trying to typecast res to a Resource. Going to leave for now to continue addressing comments but we can go back and change since it doesnt have any downstream affect

Reply:
You wouldn't need to typecast it, since the generic already specifies that it is a Resource via `T`. Might have to do `res := T(new(It))` to make it a T explicitly.